### PR TITLE
Adds the encrypt_decrypt_2 context method

### PR DESCRIPTION
**Pull request by Superhepper**

### CODE REVIEW

File: tss-esapi/tests/context_tests/tpm_commands/symmetric_primitives_tests.rs [hunk 1]
```diff
@@ -1,5 +1,146 @@
...
+        let symmetric_key_object_attributes = ObjectAttributesBuilder::new()
+            .with_user_with_auth(true)
+            .with_sign_encrypt(true)
+            .with_decrypt(true)
+            .build()
+            .expect("Failed to create object attributes for symmetric key");
+
+        let symmetric_key_public = Tpm2BPublicBuilder::new()
+            .with_type(AlgorithmIdentifier::SymCipher.into()) // This is a flaw in the builder
+            .with_name_alg(AlgorithmIdentifier::Sha256.into()) // This is a flaw in the builder.
```
Comment (ionut-arm) [inline tss-esapi/tests/context_tests/tpm_commands/symmetric_primitives_tests.rs:64]: What are the flaws here? The input type?

File: tss-esapi/src/interface_types/algorithm.rs [hunk 2]
```diff
@@ -237,6 +237,7 @@ impl TryFrom<TPMI_ALG_SYM> for SymmetricAlgorithm {
 pub enum SymmetricMode {
     Ctr,
     Ofb,
+    Cbc,
```
Comment (ionut-arm) [inline tss-esapi/src/interface_types/algorithm.rs:240]: This is definitely a breaking change - if it's in a different commit then it's alright. I'll try and think of a way to check whether the changes in a PR relate to breaking additions.

**Review by ionut-arm — COMMENTED** [review]
Looks good to me - missing docs on the new context method, but apart from that...

File: tss-esapi/tests/context_tests/tpm_commands/symmetric_primitives_tests.rs (hunk 1 above)
Comment (Superhepper) [reply tss-esapi/tests/context_tests/tpm_commands/symmetric_primitives_tests.rs:64]: Yes, it uses the raw TSS types.

**Comment by Superhepper** [comment]
@ionut-arm Thanks for your input. Now I have to figure out what I am doing wrong in the test. I get a really wierd like ```unsupported key size``` error.

**Comment by ionut-arm** [comment]
> 
> 
> @ionut-arm Thanks for your input. Now I have to figure out what I am doing wrong in the test. I get a really wierd like `unsupported key size` error.

The `unsupported key size` is in relation to the first parameter, so `TPM2B_SENSITIVE_CREATE`. If you look at your `initial_data` (which is the initial value of the key, I think), you define it as `vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 16]` which isn't 16 bytes long 😀 The same applies to a few other buffers below - you need to make them 16 bytes and then it works.

**Review by ionut-arm — APPROVED** [review]
LGTM, just a couple of typos

File: tss-esapi/src/context/tpm_commands/symmetric_primitives.rs [hunk 3]
```diff
@@ -13,8 +16,212 @@ use std::convert::TryFrom;
...
+    /// Performs symmetric encryption or decryption of the data using
+    /// the key indictaed by the `key_handle`
+    ///
+    /// # Arguments
+    /// * `key_handle` -  A [KeyHandle] to the key to be used.
+    /// * `decrypt` - A boolean indicating if the data should be decrypted or encrypted.
+    ///               If set to true the data will be decrypted else encrypted.
+    /// * `mode` - The [SymmetricMode] to be used.
+    /// * `in_data` - The data that is going to be decrypted or encrypted.
+    /// * `initial_value_in` - An initiail value as required by the algorithm.
```
[the remark this answers is not in the document]
Comment (Superhepper) [reply tss-esapi/src/context/tpm_commands/symmetric_primitives.rs]: It has been fixed

File: tss-esapi/src/context/tpm_commands/symmetric_primitives.rs [hunk 4]
```diff
@@ -13,8 +16,212 @@ use std::convert::TryFrom;
 use std::ptr::null_mut;
 
 impl Context {
-    // Missing function: EncryptDecrypt
-    // Missing function: EncryptDecrypt2
+    // Missing function: EncryptDecrypt, depricated use EncryptDecrypt2 instead.
+
+    /// Performs symmetric encryption or decryption of the data using
+    /// the key indictaed by the `key_handle`
```
[the remark this answers is not in the document]
Comment (Superhepper) [reply tss-esapi/src/context/tpm_commands/symmetric_primitives.rs]: The entire sentence was wrong and it has been fixed.

**Pull request merged**